enhancements/upgrade-readiness.md

```diff
@@ -0,0 +1,580 @@
+
+The functions provided within the library will detect if the `Probe` CRD is present on the cluster using the Discovery API. If the `Probe` resource exists, the library will provide a set of functions allowing a user to `GET` or `UPDATE` the operator's `Probe`.
+
+The library will also communicate to OLM that the operator is managing the `Probe` resource by setting the `spec.manager` field to `operator`. When the `spec.manager` field is set to `operator`, the library will throw an error if the `spec.ProbeResources` field is not `nil`.
+
+At the time of this enhancement, the operator will only need to manage its `Upgradeable` condition.
+
+#### Example: Operator Managed Probe Resource
+
+Let's consider what the `Probe` resource may look like for the Foo operator:
+
+```yaml
+apiVersion: operators.coreos.com/v1alpha1
+kind: Probe
+metadata:
+  name: foo-operator.v1.0.0
+  namespace: operators
+spec:
+  manager: operator  # When set to `operator`, the operator must manage the state of the probe resource.
+status:
+  conditions:
+  - type: Upgradeable # Overall Upgradeable condition
+    status: false     # The operator is not upgradeable
+    reason: "FooProcess"
+    message: "The operator is not upgradeable because FooProcess is running"
+```
+
+In this case, the operator is communicating that OLM should not upgrade the operator at this time. Likewise, the operator could convey that OLM may upgrade the operator with the following `Probe`:
+
+```yaml
+apiVersion: operators.coreos.com/v1alpha1
+kind: Probe
+metadata:
+  name: foo-operator.v1.0.0
+  namespace: operators
+spec:
+  manager: operator
+status:
+  conditions:
+  - type: Upgradeable # Overall Upgradeable condition
+    status: True      # The operator is upgradeable
+    reason: "ReadyForUpgrade"
+    message: "The operator is ready to be upgraded"
+```
+
+### Test Plan
+
+OLM's e2e testing suite will be upgraded to handle the usecase:
+
+- If an operator does not opt-in to "OLM Supported Conditions", OLM reverts to existing behavior.
+- If an operator opts-in to the `Upgradeable` "OLM Supported Conditions":
+  - Upgrades are not blocked if no CRs exist.
+  - Upgrades are not blocked if a CR exists but the condition is missing
+  - Upgrades are not blocked if a CR exists, the mapped condition type is present and is not equal to `false`.
+  - Upgrades are blocked if a CR exists, the mapped condition type is present and `true`.
+
+### Risks and Mitigations
+
+- OLM cannot assume that a Condition on a CR with the type `Upgradeable` maps to the `Upgradeable` "OLM Supported Condition". It is impossible for OLM to guarantee that an "OLM Supported Condition" type is not already being used as a condition type by some operator in the world. By making this feature opt-in only, OLM avoids this conflict.
```

Comment:
Well, you could scope your condition like `operatorframework.io/Upgradeable`.  That makes conflicts very unlikely, and allows you to sigh and roll your eyes dramatically if there is still a conflict ;).  But opt-in works too...

Reply:
Ha that's totally true! The goal with making things opt-in was to allow operator authors to take advantage of this feature without modifying any code / existing conditions.